core/apple/src/SinksApple.kt

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.posix.memcpy
+import platform.posix.uint8_tVar
+import kotlin.concurrent.Volatile
+
+/**
+ * Returns an output stream that writes to this sink. Closing the stream will also close this sink.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Sink.asNSOutputStream(): NSOutputStream = SinkNSOutputStream(this)
+
+@OptIn(UnsafeNumber::class)
+private class SinkNSOutputStream(
+    private val sink: Sink,
+) : NSOutputStream(toMemory = Unit) {
+
+    private val isClosed: () -> Boolean = when (sink) {
+        is RealSink -> sink::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    @Volatile
+    private var status = NSStreamStatusNotOpen
+    private var error: NSError? = null
+        set(value) {
+            status = NSStreamStatusError
+            field = value
+        }
+
+    override fun streamStatus() = status
+
+    override fun streamError() = error
+
+    override fun open() {
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpen
+        }
+    }
+
+    override fun close() {
+        status = NSStreamStatusClosed
+        sink.close()
+    }
+
+    @OptIn(DelicateIoApi::class)
+    override fun write(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+        return try {
+            if (isClosed()) throw IOException("Underlying sink is closed.")
+            if (status != NSStreamStatusOpen) return -1
+            status = NSStreamStatusWriting
+            sink.writeToInternalBuffer {
+                it.writeNative(buffer, maxLength.toInt())
+            }
+            status = NSStreamStatusOpen
+            maxLength.convert()
+        } catch (e: Exception) {
+            error = e.toNSError()
+            -1
+        }
+    }
+
+    private fun Buffer.writeNative(source: CPointer<uint8_tVar>?, maxLength: Int) {
```

Comment:
minor: I would call it something like `copyFrom` - to better reflect what we do here
Or even put `Buffer` from receiver to ordinary parameter

Reply:
I named this `writeNative()` because it does what this equivalent [`Buffer.write()` function](https://github.com/Kotlin/kotlinx-io/blob/03845eaa0f62c4d54fc5f3ea0559e0176c2c5fa5/core/common/src/Buffer.kt#L395) does, but takes a native memory pointer for its `source` parameter, rather than Kotlin's `ByteArray`. Maybe it should just be called `write()`, since it's essentially overloading this function's behavior?

Reply:
Hm, you are right! I agree
Also, I would think, that in future this 'Buffer.write(pointer)' should become public, so we can already declare it top level, but internal, but it's not really necessary, just thoughts

Reply:
That's true. This would probably fit well in a `BuffersApple` file, similar to [`BuffersJvm`](https://github.com/Kotlin/kotlinx-io/blob/03845eaa0f62c4d54fc5f3ea0559e0176c2c5fa5/core/jvm/src/BuffersJvm.kt). I'll create the file and make it `internal` for now.